### DISCUSSION TITLE
Where clause gets "ignored" in some cases when there is a limit

### DISCUSSION THREAD

**Discussion by Joyescat** [mikro-orm/mikro-orm#3363] [Q&A] [answered]
**Describe the bug**

When there is several join, a where and a limit clause in a query, the where clause gets "ignored" in some cases.

I have the following data model

*Table FileUpload*
id | companyId
-- | --
1 | 1

*Table Company*
id |
-- |
1 |

*Table User*
id | companyId|name
-- | -- | --
1 |1|John Doe
2 |1|Jane Doe

If I make a query to get all fileUpload of a specific user
```
 fileUploadRepository.createQueryBuilder(FileUpload, 'fu')
      .select(['*', 'u.name'])
      .leftJoin('fu.company', 'c')
      .leftJoin('c.users', 'u')
      .where("u.id = 1")
      .limit(10)
```

It generates the following query

```
SELECT
    `fu`.`*`,
    `u`.`name`
FROM
    `fileUpload` AS `fu`
LEFT JOIN `company` AS `ac`
ON
    `fu`.`companyId` = `c`.`id`
LEFT JOIN `user` AS `u`
ON
    `c`.`id` = `u`.`companyId`
WHERE
    `fu`.`id` IN(
    SELECT
        `fu`.`id`
    FROM
        (
        SELECT
            `fu`.`*`
        FROM
            `fileUpload` AS `fu`
        LEFT JOIN `company` AS `ac`
        ON
            `fu`.`companyId` = `c`.`id`
        LEFT JOIN `user` AS `u`
        ON
            `c`.`id` = `u`.`companyId`
        WHERE
            `u`.`id` = 1
        GROUP BY
            `fu`.`id`
        LIMIT 10
    ) AS `fu`
)
```

When there is a limit clause, mikro orm adds a subquery.
The main where clause become `WHERE fu.id IN (1)`. There is no more where on the joined table but the main query still have all the join so this can make bug.
Because, the original where clause `WHERE u.id = 1` is in the subquery

In my case, since we have 2 users we get 2 records

*Result*

fileUploadId|companyId|name
-- | -- | --
1 | 1|John Doe
1 | 1|Jane Doe

**Expected behavior**

The where clause isn’t "ignored"

To achieve that the generated query should be
```
SELECT
    `fu`.`*`
FROM
    `fileUpload` AS `fu`
WHERE
    `fu`.`id` IN(
    SELECT
        `fu`.`id`
    FROM
        (
        SELECT
            `fu`.`*`
        FROM
            `fileUpload` AS `fu`
        LEFT JOIN `company` AS `ac`
        ON
            `fu`.`companyId` = `c`.`id`
        LEFT JOIN `user` AS `u`
        ON
            `c`.`id` = `u`.`companyId`
        WHERE
            `u`.`id` = 1
        GROUP BY
            `fu`.`id`
        LIMIT 10
    ) AS `fu`
)
```
Since we don’t need all the join on the main query, because the where clause of the main query only involve the table we select from. In our example `fileUpload`

I don’t know if this is the right way to fix it but if we you can make it without the subquery it can even be easier
```
SELECT
    `fu`.`*`
FROM
    `fileUpload` AS `fu`
LEFT JOIN `company` AS `ac`
ON
    `fu`.`companyId` = `c`.`id`
LEFT JOIN `user` AS `u`
ON
    `c`.`id` = `u`.`companyId`
WHERE
    `u`.`id` = 1
GROUP BY
    `fu`.`id`
LIMIT 10
```

Dependency | Version
-- | --
node | 14.17.5
typescript | 4.4.4
mikro-orm | 5.2.1

**Comment by B4nan** [comment, marked as answer]
This is on purpose, while it may look weird for direct QB usage but it is needed internally - as we want to prune only the root entity, not the joined relations. I guess you are doing some `qb.execute()` as this shouldnt be a problem with `qb.getResult()` - you havent really shared what you are doing, only QB definition, not code that would actually evaluate it...), you can disable automatic pagination on QB via `qb.setFlag(QueryFlag.DISABLE_PAGINATE)` (its enabled automatically if you use a query with limit and to-many join).

We might change the defaults in v6 a bit, but I dont think we want to change how the query is built.

> Since we don’t need all the join on the main query

You are the one who joins the tables and selects the fields from them - you select the user name. Your query is not really equivalent to what you instruct the QB.

**Reply by Joyescat** [reply]
My bad, I forgot we need the join for the select and not only the where clause.

Using the flag `DISABLE_PAGINATE` fix it.

Thanks for your help !
